### PR TITLE
Seg Mesh Extracting and Seg Extracting from GarfVDB

### PR DESCRIPTION
**Pull request by zlalena**
Cleaned up versions of the code used for the GTC SJ lab

1. File to extract mesh segemnets from a mesh made from teh full orginal GS scene\
- Added a couple things to fill in the giant gaps at the bottom of meshes
- tested on safety park (outdoor airborn imagery)
3. Mass segment extraction into plys using garfvdb

### CODE REVIEW

File: instance_segmentation/garfvdb/extract_segments.py [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+from pathlib import Path
+from typing import Literal
+
+import numpy as np
+import torch
+from fvdb import GaussianSplat3d
+from scipy.spatial import cKDTree
+
+# Import directly to avoid fvdb_reality_capture.tools.__init__ pulling optional deps (e.g. DLNR).
+from fvdb_reality_capture.tools._filter_splats import (
```
Comment (swahtz) [inline instance_segmentation/garfvdb/extract_segments.py]: **🚩 issue:** ‏We shouldn't import through the internal `_` modules.  These examples are meant to demonstrate best practices to users and if DLNR is causing an issue in `fvdb_reality_capture`, we should fix that and not try to work around it in the examples.

File: instance_segmentation/garfvdb/extract_segments.py [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+
+    parser.add_argument("--n", type=int, default=10, help="Number of segments to export")
+    parser.add_argument(
+        "--scale",
+        type=float,
+        default=0.1,
+        help="Clustering scale (absolute or fraction of max; see --scale-is-fraction-of-max)",
+    )
+    parser.add_argument(
+        "--scale-is-fraction-of-max",
```
Comment (swahtz) [inline instance_segmentation/garfvdb/extract_segments.py]: **🎵 note (non-blocking):** ‏This option is a bit confusing, I think this should just be expressed as an option that turns off this behaviour.  My understanding is that by default, the value of this option is `True` and the scale value is a fraction of the `max_grouping_scale` (in 'world' units), but this option when set to false (`--scale-is-fraction-of-max=False`?), turns the interpretation of scale into a world-unit measurement.

I think it makes more sense to just have this option be something like `--scale-is-world-unit` (by default it would be off) to turn on interpreting the scale value as the world unit and not a fraction.

File: instance_segmentation/garfvdb/extract_segments.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+    )
+    parser.add_argument("--seed", type=int, default=42, help="Random seed for sampling/subsampling")
+    parser.add_argument("--device", type=str, default="cuda", help="Torch device")
+    parser.add_argument("--verbose", action="store_true", help="Enable debug logging")
+
+    parser.add_argument(
+        "--min-splat-scale",
+        type=float,
+        default=0.1,
+        help="Drop Gaussians with scale below this value (0 disables)",
```
Comment (swahtz) [inline instance_segmentation/garfvdb/extract_segments.py]: **💡 suggestion:**  I think it's worth mentioning whether the minimum scale here is in world units or a similar fraction of the max scale as above.‏

File: instance_segmentation/garfvdb/extract_segments.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+        nargs="*",
+        default=[0.96, 0.96, 0.96, 0.96, 0.98, 0.99],
+        help="Per-channel mean percentiles for splat pre-filtering",
+    )
+
+    parser.add_argument(
+        "--min-cluster-gaussians",
+        type=int,
+        default=200,
+        help="Drop clusters smaller than this (0 disables)",
```
Comment (swahtz) [inline instance_segmentation/garfvdb/extract_segments.py]: **💡 suggestion:** ‏'smaller' is a bit confusing because it can be read as a 3D size/scale.  I think it probably makes more sense to say "Drop clusters with a number of member Gaussians less than this value"

File: instance_segmentation/garfvdb/extract_segments.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+
+    max_scale = float(segmentation_model.max_grouping_scale.item())
+    scale_abs = float(scale) * max_scale if scale_is_fraction_of_max else float(scale)
+    logger.info("Segmentation model max scale: %.6f", max_scale)
+    logger.info("Clustering at scale: %.6f", scale_abs)
+
+    clustering_gs_model = gs_model
+    subsample_mask: torch.Tensor | None = None
+    if max_gaussians_for_clustering > 0 and gs_model.num_gaussians > max_gaussians_for_clustering:
+        logger.warning(
```
Comment (swahtz) [inline instance_segmentation/garfvdb/extract_segments.py]: **🤓 nitpick (non-blocking):** ‏I think this should just be an 'info' message like the one below, I don't believe this is warning about anything

File: instance_segmentation/garfvdb/extract_segments.py [hunk 6]
```diff
@@ -0,0 +1,466 @@
+# Copyright Contributors to the OpenVDB Project
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Export instance segments from a trained GARfVDB segmentation checkpoint.
+
+Loads a Gaussian splat reconstruction and segmentation checkpoint, clusters
+per-Gaussian affinity features at a chosen scale, filters clusters, then writes
+``n`` segment ``.ply`` files (one per selected cluster).
```
Comment (swahtz) [inline instance_segmentation/garfvdb/extract_segments.py]: **🤓 nitpick (non-blocking):** ‏I think it's worth mentioning these `ply` files are Gaussian splat scenes as well, so they're not going to be confused with ply meshes.
 
```suggestion
``n`` segment ``.ply`` Gaussian splat scenes (one per selected cluster).
```

File: segmenation_extraction/extract_mesh_for_segments.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    vertices = smoothed
+
+    return vertices.astype(np.float32), faces
+
+
+def main() -> None:
+    logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
+
+    parser = argparse.ArgumentParser(
+        description="Segment full scene mesh based on a PLY Gaussian Splat segment",
```
Comment (swahtz) [inline segmenation_extraction/extract_mesh_for_segments.py]: **🚩 issue:** ‏I'm confused about what this script does based on this description and the documentation at the top of the file.  The description should be improved so users understand what this script will do.

File: segment_mesh_extraction/extract_mesh_segments.py
Comment (swahtz) [inline segment_mesh_extraction/extract_mesh_segments.py:1]: **💡 suggestion:** ‏If this mesh extraction is going to be directly related to the segment extraction in GARfVDB, then I think we should just move this file to the instance_segementation/garfvdb directory.  Otherwise, we should include a README, environment files, pyproject.toml, etc. in this directory for a user to be able to run this project.

File: segment_mesh_extraction/extract_mesh_segments.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+Optionally closes holes via harmonic Laplacian fill then make_mesh_watertight
+Output can be used as input for fvdb-reality-capture/scripts/create_isaac_ready_files.py to make a USDZ
+"""
+from __future__ import annotations
+
+import argparse
+import logging
+from pathlib import Path
+
+import igl
```
Comment (swahtz) [inline segment_mesh_extraction/extract_mesh_segments.py:25]: **☑️ todo:** ‏We should include igl and point_cloud_utils in GARfVDB's environment file so these modules are available in the example.

File: segmenation_extraction/extract_mesh_for_segments.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+    _, indices = tree.query(query_vertices, k=1, workers=-1)
+    return source_colors[indices]
+
+
+def _has_vertex_colors(vertex_colors: np.ndarray | None) -> bool:
+    """Return True if the mesh has a non-empty per-vertex color array."""
+    return vertex_colors is not None and vertex_colors.size > 0 and vertex_colors.shape[0] > 0
+
+
+def _normalize_vertex_colors(colors: np.ndarray) -> np.ndarray:
```
Comment (swahtz) [inline segmenation_extraction/extract_mesh_for_segments.py]:
```suggestion
def _as_float_vertex_colors(colors: np.ndarray) -> np.ndarray:
```

This isn't really normalization in the statistical/vector sense, I'd suggest we change the name to something like `_as_float_vertex_colors` or `_as_unit_float_vertex_colors` so it's clear we're not changing the meaning in any way, we're just scaling to change formats.

File: segmenation_extraction/extract_mesh_for_segments.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    # Save the extracted mesh
+    output_path.parent.mkdir(parents=True, exist_ok=True)
+    logger.info(f"Saving extracted mesh to {output_path}")
+
+    if extracted_colors is not None:
+        pcu.save_mesh_vfc(
+            str(output_path),
+            extracted_vertices,
+            extracted_faces_reindexed,
+            _normalize_vertex_colors(extracted_colors),
```
Comment (swahtz) [inline segmenation_extraction/extract_mesh_for_segments.py]:
```suggestion
            extracted_colors,
```


**☑️ todo:** ‏We already called this function on the colors at line 154, I think we can just avoid calling this again.

**Review by swahtz — CHANGES_REQUESTED** [review]
Overall looks pretty good, just a few organizational changes that are important and documentation/user experience changes to address.

File: segment_mesh_extraction/extract_mesh_segments.py (hunk 8 above)
Comment (zlalena) [reply segment_mesh_extraction/extract_mesh_segments.py:25]: added in segment mesh extraction

File: instance_segmentation/garfvdb/extract_segments.py (hunk 2 above)
Comment (zlalena) [reply instance_segmentation/garfvdb/extract_segments.py]: chnaged to hoefully be more easy to understand

Comment (swahtz) [reply instance_segmentation/garfvdb/extract_segments.py]: Thanks, that's great

File: segmenation_extraction/extract_mesh_for_segments.py (hunk 7 above)
Comment (swahtz) [reply segmenation_extraction/extract_mesh_for_segments.py]: Thanks for resolving that

File: segment_mesh_extraction/extract_mesh_segments.py
Comment (swahtz) [reply segment_mesh_extraction/extract_mesh_segments.py:1]: Great, thanks for setting up the project files

**Review by swahtz — APPROVED** [review]
Thanks for addressing the notes!
